File: archive/task31/task31.py
from typing import List
import unittest


class Interval(object):
    def __init__(self, start, end):
        self.start = start
        self.end = end


class Solution:
    """
    @param intervals: an array of meeting time intervals
    @return: if a person could attend all meetings
    """

    def can_attend_meetings(self, intervals: List[Interval]) -> bool:
        intervals = sorted(intervals, key=lambda i: (i.start, i.end))

        for i1, i2 in zip(intervals, intervals[1:]):
            if i2.start < i1.end:
                return False

        return True


class TestMethods(unittest.TestCase):

    def test1(self):
        s = Solution()

        times = [
            Interval(0, 30),
            Interval(5, 10),
            Interval(15, 20),
        ]

        self.assertEqual(False, s.can_attend_meetings(times))

    def test2(self):
        s = Solution()

        times = [
            Interval(0, 8),
            Interval(8, 10),
        ]

        self.assertEqual(True, s.can_attend_meetings(times))

    def test3(self):
        s = Solution()

        times = [
            Interval(5, 8),
            Interval(9, 15),
        ]

        self.assertEqual(True, s.can_attend_meetings(times))


if __name__ == '__main__':
    unittest.main()

File: archive/task31/task31.cpp
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include "../doctest.h"

#include <vector>
#include <vector>

/**
 *
 * Given an array of meeting time intervals consisting of start and end times [[s1,e1],[s2,e2],...] (si < ei), find
 * the minimum number of conference rooms required.
 * For example,
 * Given [[0, 30],[5, 10],[15, 20]],
 * return 2.
 *
 */

class Interval
{
public:
    int start, end;
    Interval(int start, int end)
    {
        this->start = start;
        this->end = end;
    }
};

class Solution
{
private:
    static bool compareByLength(const Interval &i1, const Interval &i2)
    {
        return i1.start < i2.start || i1.start == i2.start && i1.end < i2.end;
    }

public:
    /**
     * @param intervals: an array of meeting time intervals
     * @return: if a person could attend all meetings
     */
    bool canAttendMeetings(std::vector<Interval> &intervals)
    {
        std::sort(intervals.begin(), intervals.end(), compareByLength);

        size_t len_intervals = intervals.size();

        for (size_t i = 1; i < len_intervals; ++i)
        {
            if (intervals[i].start < intervals[i - 1].end)
            {
                return false;
            }
        }

        return true;
    }
};

// Time complexity: O(nlogn)
// Space complexity: O(1)

TEST_CASE("Example 1")
{
    Solution s;

    std::vector<Interval> times = {
        Interval(0, 30),
        Interval(5, 10),
        Interval(15, 20),
    };

    unsigned result = s.canAttendMeetings(times);
    unsigned expected = false;

    CHECK_EQ(result, expected);
}

TEST_CASE("Example 2")
{
    Solution s;

    std::vector<Interval> times = {
        Interval(0, 8),
        Interval(8, 10),
    };

    unsigned result = s.canAttendMeetings(times);
    unsigned expected = true;

    CHECK_EQ(result, expected);
}

TEST_CASE("Example 3")
{
    Solution s;

    std::vector<Interval> times = {
        Interval(5, 8),
        Interval(9, 15),
    };

    unsigned result = s.canAttendMeetings(times);
    unsigned expected = true;

    CHECK_EQ(result, expected);
}
